core/templates/pages/exploration-editor-page/services/version-history.service.ts

```diff
@@ -0,0 +1,314 @@
+      oldVersionNumber: (
+        this.fetchedStateVersionNumbers[
+          this.currentPositionInStateVersionHistoryList - 1]),
+      newVersionNumber: (
+        this.fetchedStateVersionNumbers[
+          this.currentPositionInStateVersionHistoryList - 2]),
+      committerUsername: (
+        this.fetchedCommitterUsernames[
+          this.currentPositionInStateVersionHistoryList - 1])
+    };
+  }
+
+  canShowBackwardMetadataDiffData(): boolean {
+    return (
+      this.currentPositionInMetadataVersionHistoryList >= 0 &&
+      this.currentPositionInMetadataVersionHistoryList <
+        this.fetchedMetadataVersionNumbers.length - 1 &&
+      this.fetchedMetadataVersionNumbers[
+        this.currentPositionInMetadataVersionHistoryList + 1] !== null &&
+      this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList + 1] !== null
+    );
+  }
+
+  canShowForwardMetadataDiffData(): boolean {
+    return (
+      this.currentPositionInMetadataVersionHistoryList >= 2 &&
+      this.currentPositionInMetadataVersionHistoryList <
+        this.fetchedMetadataVersionNumbers.length &&
+      this.fetchedMetadataVersionNumbers[
+        this.currentPositionInMetadataVersionHistoryList - 1] !== null &&
+      this.fetchedMetadataVersionNumbers[
+        this.currentPositionInMetadataVersionHistoryList - 2] !== null &&
+      this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList - 1] !== null &&
+      this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList - 2] !== null
+    );
+  }
+
+  getBackwardMetadataDiffData(): MetadataDiffData {
+    return {
+      oldMetadata: this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList + 1],
+      newMetadata: this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList],
+      oldVersionNumber: (
+        this.fetchedMetadataVersionNumbers[
+          this.currentPositionInMetadataVersionHistoryList + 1]),
+      newVersionNumber: (
+        this.fetchedMetadataVersionNumbers[
+          this.currentPositionInMetadataVersionHistoryList]),
+      committerUsername: (
+        this.fetchedCommitterUsernames[
+          this.currentPositionInMetadataVersionHistoryList + 1])
+    };
+  }
+
+  getForwardMetadataDiffData(): MetadataDiffData {
+    return {
+      oldMetadata: this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList - 1],
+      newMetadata: this.fetchedMetadata[
+        this.currentPositionInMetadataVersionHistoryList - 2],
+      oldVersionNumber: (
+        this.fetchedMetadataVersionNumbers[
+          this.currentPositionInMetadataVersionHistoryList - 1]),
+      newVersionNumber: (
+        this.fetchedMetadataVersionNumbers[
+          this.currentPositionInMetadataVersionHistoryList - 2]),
+      committerUsername: (
+        this.fetchedCommitterUsernames[
+          this.currentPositionInMetadataVersionHistoryList - 1])
```

Comment:
Do we need to have checks / guard conditions to ensure that undefined isn't returned when we access an index that is out of bounds?

Reply:
This cannot happen as before fetching these data, we call the functions such as canShowBackwardMetadataDiffData and canShowForwardMetadataDiffData which act as the guard conditions.